// ngx-extended-pdf-viewer/esm2022/lib/toolbar/pdf-stamp-editor/pdf-stamp-editor.component.mjs
import { Component, Input, effect } from '@angular/core';
import { PositioningService } from '../../dynamic-css/positioning.service';
import { AnnotationEditorType } from '../../options/editor-annotations';
import { getVersionSuffix, pdfDefaultOptions } from '../../options/pdf-default-options';
import * as i0 from "@angular/core";
import * as i1 from "../../pdf-notification-service";
import * as i2 from "../../dynamic-css/positioning.service";
import * as i3 from "../../focus-management.service";
import * as i4 from "../pdf-shy-button/pdf-shy-button.component";
import * as i5 from "../../responsive-visibility";
export class PdfStampEditorComponent {
    cdr;
    positioningService;
    focusManagement;
    show = true;
    isSelected = false;
    PDFViewerApplication;
    get pdfJsVersion() {
        return getVersionSuffix(pdfDefaultOptions.assetsFolder);
    }
    constructor(notificationService, cdr, positioningService, focusManagement) {
        this.cdr = cdr;
        this.positioningService = positioningService;
        this.focusManagement = focusManagement;
        effect(() => {
            this.PDFViewerApplication = notificationService.onPDFJSInitSignal();
            if (this.PDFViewerApplication) {
                this.onPdfJsInit();
            }
        });
    }
    onPdfJsInit() {
        this.PDFViewerApplication?.eventBus.on('annotationeditormodechanged', ({ mode }) => {
            setTimeout(() => {
                const wasSelected = this.isSelected;
                this.isSelected = mode === 13;
                // Focus management
                if (!wasSelected && this.isSelected) {
                    // Dialog just opened
                    this.focusManagement.moveFocusToDialog('editorStampParamsToolbar', 'Stamp editor toolbar opened', 'primaryEditorStamp');
                }
                else if (wasSelected && !this.isSelected) {
                    // Dialog just closed
                    this.focusManagement.returnFocusToPrevious('Stamp editor toolbar closed');
                }
                this.cdr.detectChanges();
            });
        });
    }
    onClick(event) {
        const currentMode = this.PDFViewerApplication?.pdfViewer.annotationEditorMode;
        this.PDFViewerApplication?.eventBus.dispatch('switchannotationeditormode', {
            source: this,
            mode: currentMode === AnnotationEditorType.STAMP ? AnnotationEditorType.NONE : AnnotationEditorType.STAMP,
            isFromKeyboard: event.detail === 0,
        });
        const positioningService = new PositioningService();
        positioningService.positionPopupBelowItsButton('primaryEditorStamp', 'editorStampParamsToolbar');
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: PdfStampEditorComponent, deps: [{ token: i1.PDFNotificationService }, { token: i0.ChangeDetectorRef }, { token: i2.PositioningService }, { token: i3.FocusManagementService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.12", type: PdfStampEditorComponent, selector: "pdf-stamp-editor", inputs: { show: "show" }, ngImport: i0, template: "<pdf-shy-button title=\"Text\" primaryToolbarId=\"primaryEditorStamp\"\n  [cssClass]=\"show | responsiveCSSClass : 'hiddenTinyView'\" l10nId=\"pdfjs-editor-stamp-button\"\n  l10nLabel=\"pdfjs-editor-stamp-button-label\" role=\"radio\" ariaHasPopup=\"true\"\n  ariaControls=\"editorStampParamsToolbar\" [order]=\"4200\" [action]=\"onClick\" [toggled]=\"isSelected\"\n  [closeOnClick]=\"true\"\n  image=\"<svg aria-hidden='true' focusable='false' width='20px' height='20px' viewBox='0 0 24 24'> <path fill='currentColor' d='M13 19C13 19.7 13.13 20.37 13.35 21H5C3.9 21 3 20.11 3 19V5C3 3.9 3.9 3 5 3H19C20.11 3 21 3.9 21 5V13.35C20.37 13.13 19.7 13 19 13V5H5V19H13M13.96 12.29L11.21 15.83L9.25 13.47L6.5 17H13.35C13.75 15.88 14.47 14.91 15.4 14.21L13.96 12.29M20 18V15H18V18H15V20H18V23H20V20H23V18H20Z' /> </svg>\">\n</pdf-shy-button>", styles: ["button{padding:0;height:25px;background-color:transparent;width:100%}button:focus{outline:none;border:none}.align-image-to-text{top:3px;position:relative;padding-right:4px}\n"], dependencies: [{ kind: "component", type: i4.PdfShyButtonComponent, selector: "pdf-shy-button", inputs: ["primaryToolbarId", "secondaryMenuId", "cssClass", "eventBusName", "l10nId", "l10nLabel", "title", "toggled", "disabled", "order", "action", "closeOnClick", "onlySecondaryMenu", "ariaHasPopup", "ariaControls", "role", "image"] }, { kind: "pipe", type: i5.ResponsiveCSSClassPipe, name: "responsiveCSSClass" }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: PdfStampEditorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pdf-stamp-editor', template: "<pdf-shy-button title=\"Text\" primaryToolbarId=\"primaryEditorStamp\"\n  [cssClass]=\"show | responsiveCSSClass : 'hiddenTinyView'\" l10nId=\"pdfjs-editor-stamp-button\"\n  l10nLabel=\"pdfjs-editor-stamp-button-label\" role=\"radio\" ariaHasPopup=\"true\"\n  ariaControls=\"editorStampParamsToolbar\" [order]=\"4200\" [action]=\"onClick\" [toggled]=\"isSelected\"\n  [closeOnClick]=\"true\"\n  image=\"<svg aria-hidden='true' focusable='false' width='20px' height='20px' viewBox='0 0 24 24'> <path fill='currentColor' d='M13 19C13 19.7 13.13 20.37 13.35 21H5C3.9 21 3 20.11 3 19V5C3 3.9 3.9 3 5 3H19C20.11 3 21 3.9 21 5V13.35C20.37 13.13 19.7 13 19 13V5H5V19H13M13.96 12.29L11.21 15.83L9.25 13.47L6.5 17H13.35C13.75 15.88 14.47 14.91 15.4 14.21L13.96 12.29M20 18V15H18V18H15V20H18V23H20V20H23V18H20Z' /> </svg>\">\n</pdf-shy-button>", styles: ["button{padding:0;height:25px;background-color:transparent;width:100%}button:focus{outline:none;border:none}.align-image-to-text{top:3px;position:relative;padding-right:4px}\n"] }]
        }], ctorParameters: () => [{ type: i1.PDFNotificationService }, { type: i0.ChangeDetectorRef }, { type: i2.PositioningService }, { type: i3.FocusManagementService }], propDecorators: { show: [{
                type: Input
            }] } });
