// ngx-extended-pdf-viewer/esm2022/lib/pdf-csp-policy.service.mjs
import { CSP_NONCE, Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PdfCspPolicyService {
    sanitizer = undefined; // TrustedTypePolicy;
    csp_nonce;
    constructor() { }
    init() {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        if (this.sanitizer) {
            // already initialized
            return;
        }
        const ttWindow = globalThis;
        if (ttWindow.trustedTypes) {
            this.sanitizer = ttWindow.trustedTypes.createPolicy('pdf-viewer', {
                createHTML: (input) => input,
                createScriptURL: (input) => input,
            });
        }
    }
    addTrustedCSS(styles, css) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        this.init();
        if (this.sanitizer) {
            styles.textContent = this.sanitizer.createHTML(css);
        }
        else {
            styles.textContent = css;
        }
    }
    addTrustedJavaScript(scripts, css) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        this.init();
        if (this.sanitizer) {
            scripts.src = this.sanitizer.createScriptURL(css);
        }
        else {
            scripts.src = css;
        }
    }
    sanitizeHTML(html) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return '';
        }
        this.init();
        if (this.sanitizer) {
            return this.sanitizer.createHTML(html);
        }
        else {
            return html;
        }
    }
    addTrustedHTML(element, html) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        this.init();
        if (this.sanitizer) {
            element.innerHTML = this.sanitizer.createHTML(html);
        }
        else {
            element.innerHTML = html;
        }
    }
    createTrustedHTML(html) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        this.init();
        if (this.sanitizer) {
            return this.sanitizer.createHTML(html);
        }
        else {
            return html;
        }
    }
    generateTrustedURL(sourcePath) {
        if (typeof window === 'undefined') {
            // server-side rendering
            return;
        }
        this.init();
        if (this.sanitizer) {
            return this.sanitizer.createScriptURL(sourcePath);
        }
        return sourcePath;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: PdfCspPolicyService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: PdfCspPolicyService, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: PdfCspPolicyService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [], propDecorators: { csp_nonce: [{
                type: Inject,
                args: [CSP_NONCE]
            }] } });
