// primeng/fesm2022/primeng-classnames.mjs
import * as i0 from '@angular/core';
import { input, computed, Directive, NgModule } from '@angular/core';
import { cn } from '@primeuix/utils';

/**
 * PClass directive provides extends class binding functionality.
 * Supports strings, arrays, objects, and mixed combinations.
 * @group Components
 */
class ClassNames {
    /**
     * Class value(s) to be applied. Can be a string, array, object, or combination.
     * @group Props
     */
    classNames = input(undefined, { ...(ngDevMode ? { debugName: "classNames" } : {}), alias: 'pClass' });
    classes = computed(() => cn(this.classNames()), ...(ngDevMode ? [{ debugName: "classes" }] : []));
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.0.5", ngImport: i0, type: ClassNames, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "21.0.5", type: ClassNames, isStandalone: true, selector: "[pClass]", inputs: { classNames: { classPropertyName: "classNames", publicName: "pClass", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "classes()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.0.5", ngImport: i0, type: ClassNames, decorators: [{
            type: Directive,
            args: [{
                    selector: '[pClass]',
                    standalone: true,
                    host: {
                        '[class]': 'classes()'
                    }
                }]
        }], propDecorators: { classNames: [{ type: i0.Input, args: [{ isSignal: true, alias: "pClass", required: false }] }] } });
class ClassNamesModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.0.5", ngImport: i0, type: ClassNamesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "21.0.5", ngImport: i0, type: ClassNamesModule, imports: [ClassNames], exports: [ClassNames] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "21.0.5", ngImport: i0, type: ClassNamesModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.0.5", ngImport: i0, type: ClassNamesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [ClassNames],
                    exports: [ClassNames]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { ClassNames, ClassNamesModule };
