// ngx-extended-pdf-viewer/esm2022/lib/dynamic-css/positioning.service.mjs
import { pdfDefaultOptions } from '../options/pdf-default-options';
export class PositioningService {
    static DOORHANGER_OFFSET = 17;
    static TOOLBAR_MARGIN = 4;
    positionPopupBelowItsButton(buttonId, popupId) {
        if (!pdfDefaultOptions.positionPopupDialogsWithJavaScript) {
            return;
        }
        setTimeout(() => {
            const button = this.findVisibleButton(buttonId);
            const popup = this.getPopupElement(popupId);
            if (!button || !popup)
                return;
            this.applyPopupPositioning(button, popup);
        });
    }
    findVisibleButton(buttonId) {
        const findVisible = (selector) => Array.from(document.querySelectorAll(`#${selector}`)).find((el) => el.offsetParent !== null);
        return findVisible(buttonId) || findVisible('secondaryToolbarToggle') || null;
    }
    getPopupElement(popupId) {
        const popup = document.querySelector(`#${popupId}`);
        return popup?.offsetParent ? popup : null;
    }
    applyPopupPositioning(button, popup) {
        const popupContainer = popup.offsetParent;
        const buttonRect = button.getBoundingClientRect();
        const containerRect = popupContainer.getBoundingClientRect();
        const isRTL = document.documentElement.getAttribute('dir') === 'rtl';
        let isRightAligned = isRTL;
        if (popup.classList.contains('doorHangerRight')) {
            isRightAligned = !isRTL;
        }
        this.setBasicPopupStyles(popup, isRightAligned);
        this.setHorizontalPosition(popup, buttonRect, containerRect, isRightAligned);
        this.setVerticalPosition(popup, containerRect);
    }
    setBasicPopupStyles(popup, isRightAligned) {
        Object.assign(popup.style, {
            position: 'absolute',
            display: 'block',
            transformOrigin: isRightAligned ? 'top right' : 'top left',
        });
    }
    setHorizontalPosition(popup, buttonRect, containerRect, isRightAligned) {
        const centerOffset = buttonRect.width / 2 - PositioningService.DOORHANGER_OFFSET;
        if (!isRightAligned) {
            popup.style.left = `${buttonRect.left - containerRect.left + centerOffset}px`;
            popup.style.right = '';
        }
        else {
            popup.style.right = `${containerRect.right - buttonRect.right + centerOffset}px`;
            popup.style.left = '';
        }
    }
    setVerticalPosition(popup, containerRect) {
        const toolbarContainer = document.querySelector('#toolbarContainer');
        if (!toolbarContainer)
            return;
        const toolbarRect = toolbarContainer.getBoundingClientRect();
        const top = toolbarRect.bottom - containerRect.top + PositioningService.TOOLBAR_MARGIN;
        popup.style.top = `${top}px`;
    }
}
