// ngx-extended-pdf-viewer/esm2022/lib/options/editor-annotations.mjs
export var AnnotationMode;
(function (AnnotationMode) {
    AnnotationMode[AnnotationMode["DISABLE"] = 0] = "DISABLE";
    AnnotationMode[AnnotationMode["ENABLE"] = 1] = "ENABLE";
    AnnotationMode[AnnotationMode["ENABLE_FORMS"] = 2] = "ENABLE_FORMS";
    AnnotationMode[AnnotationMode["ENABLE_STORAGE"] = 3] = "ENABLE_STORAGE";
})(AnnotationMode || (AnnotationMode = {}));
export var AnnotationEditorType;
(function (AnnotationEditorType) {
    AnnotationEditorType[AnnotationEditorType["DISABLE"] = -1] = "DISABLE";
    AnnotationEditorType[AnnotationEditorType["NONE"] = 0] = "NONE";
    AnnotationEditorType[AnnotationEditorType["FREETEXT"] = 3] = "FREETEXT";
    AnnotationEditorType[AnnotationEditorType["HIGHLIGHT"] = 9] = "HIGHLIGHT";
    AnnotationEditorType[AnnotationEditorType["STAMP"] = 13] = "STAMP";
    AnnotationEditorType[AnnotationEditorType["INK"] = 15] = "INK";
    AnnotationEditorType[AnnotationEditorType["POPUP"] = 16] = "POPUP";
    AnnotationEditorType[AnnotationEditorType["SIGNATURE"] = 101] = "SIGNATURE";
    AnnotationEditorType[AnnotationEditorType["COMMENT"] = 102] = "COMMENT";
})(AnnotationEditorType || (AnnotationEditorType = {}));
export const AnnotationEditorParamsType = {
    RESIZE: 1,
    CREATE: 2,
    FREETEXT_SIZE: 11,
    FREETEXT_COLOR: 12,
    FREETEXT_OPACITY: 13,
    INK_COLOR: 21,
    INK_THICKNESS: 22,
    INK_OPACITY: 23,
    HIGHLIGHT_COLOR: 31,
    HIGHLIGHT_DEFAULT_COLOR: 32,
    HIGHLIGHT_THICKNESS: 33,
    HIGHLIGHT_FREE: 34,
    HIGHLIGHT_SHOW_ALL: 35,
    DRAW_STEP: 41,
};
